tests/validator/validator.cpp

```diff
@@ -2841,3 +2956,465 @@ TEST(Validator, duplicateIdAll)
+        + resourcePath("importer/") + "circularImport_1.cellml'; and\n"
+                                      " - component 'i_am_cyclic' references component 'c2' in '"
+        + resourcePath("importer/") + "circularImport_2.cellml'.";
+
+    auto parser = libcellml::Parser::create();
+    auto validator = libcellml::Validator::create();
+    auto importer = libcellml::Importer::create();
+
+    auto model = parser->parseModel(fileContents("importer/circularImport_1_duplicated_name.cellml"));
+    EXPECT_EQ(size_t(0), parser->issueCount());
+
+    validator->validateModel(model);
+    EXPECT_EQ(size_t(1), validator->issueCount());
+    EXPECT_EQ(errorMessagesValidator[1], validator->issue(0)->description());
+    EXPECT_EQ(model, validator->error(0)->model());
+
+    importer->resolveImports(model, resourcePath("importer/"));
+    EXPECT_EQ(size_t(1), importer->errorCount());
+    EXPECT_EQ(errorMessageImporter, importer->issue(0)->description());
+    EXPECT_EQ(model->units(0), importer->issue(0)->units());
+
+    validator->validateModel(model);
+    EXPECT_EQ(size_t(2), validator->issueCount());
+    EXPECT_EQ_ISSUES(errorMessagesValidator, validator);
+}
+
+TEST(Validator, importComponentWithInvalidName)
+{
+    const std::vector<std::string> errorMessages = {
+        "Import of component 'c' has an invalid URI in the xlink:href attribute.",
+        "The attempt to resolve imports with the model at '"
+            + resourcePath("importer/")
+            + "i am broken and invalid.cellml' failed: the file could not be opened.",
+    };
+
+    auto parser = libcellml::Parser::create();
+    auto validator = libcellml::Validator::create();
+    auto importer = libcellml::Importer::create();
+
+    auto model = parser->parseModel(fileContents("importer/invalid_import_url.cellml"));
+    EXPECT_EQ(size_t(0), parser->issueCount());
+
+    validator->validateModel(model);
+    EXPECT_EQ(size_t(1), validator->issueCount());
+    EXPECT_EQ(errorMessages[0], validator->issue(0)->description());
+    EXPECT_EQ(model->component("c")->importSource(), validator->error(0)->importSource());
+
+    importer->resolveImports(model, resourcePath("importer/"));
+    EXPECT_EQ(size_t(1), importer->errorCount());
+    EXPECT_EQ(errorMessages[1], importer->issue(0)->description());
+
+    validator->validateModel(model);
+    EXPECT_EQ(size_t(1), validator->issueCount());
+    EXPECT_EQ(errorMessages[0], validator->issue(0)->description());
+    EXPECT_EQ(model->component("c")->importSource(), validator->error(0)->importSource());
+}
+
+TEST(Validator, importSecondGenComponentWithInvalidName)
+{
+    const std::string errorMessageValidator =
+        "Imported component 'c' is not valid because:\n"
+        "  -> Component 'c' imported as 'c' from 'invalid_import_url.cellml' has an error:\n"
+        "  -> Import of component 'c' has an invalid URI in the xlink:href attribute.";
```

Comment:
The use of "because", arrows, capital letters, "has an error:", etc. breaks our current convention and, in fact, makes this error message very confusing. At this stage, and assuming I understand the error message right, I would probably have something like:
```
Component 'c' is imported from component 'c' in 'invalid_import_url.cellml' which is invalid (it has an invalid URI in the 'xlink:href' attribute).
```

Reply:
I agree (partly).  The reason for it is that you don't know the depth of the error.  It could be four or five layers of imports deep.  I think we need to discuss this reporting of imports errors in more depth.

Reply:
Agreed, something that needs to be discussed.